rust/src/stun/stun.rs

```diff
@@ -0,0 +1,511 @@
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&StunTransaction> {
+        self.transactions.iter().find(|&tx| tx.tx_id == tx_id + 1)
+    }
+
+    fn new_tx(&mut self) -> StunTransaction {
+        let mut tx = StunTransaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        return tx;
+    }
+
+    fn find_request(&mut self) -> Option<&mut StunTransaction> {
+        self.transactions.iter_mut().find(|tx| tx.response.is_none())
+    }
+
+    fn parse_request(&mut self, input: &[u8]) -> AppLayerResult {
+        // We're not interested in empty requests.
+        if input.is_empty() {
+            return AppLayerResult::ok();
+        }
+
+        // If there was gap, check we can sync up again.
+        if self.request_gap {
+            if !probe(input) {
+                // The parser now needs to decide what to do as we are not in sync.
+                // For this stun, we'll just try again next time.
+                return AppLayerResult::ok();
+            }
+
+            // It looks like we're in sync with a message header, clear gap
+            // state and keep parsing.
+            self.request_gap = false;
+        }
+
+        let mut start = input;
+        while !start.is_empty() {
+            match parser::parse_message(start) {
+                Ok((rem, request)) => {
+                    start = rem;
+
+                    SCLogDebug!("Request: {:?}", request);
+                    let mut tx = self.new_tx();
+                    tx.request = Some(request);
+                    self.transactions.push_back(tx);
+                },
+                Err(Err::Incomplete(_)) => {
+                    // TODO create events
+                    // UDP doesn't allow for incomplete segments, so we return error
+                    SCLogDebug!("Incomplete request.");
+                    return AppLayerResult::err();
+                },
+                Err(_) => {
+                    return AppLayerResult::err();
+                },
+            }
+        }
+
+        // Input was fully consumed.
+        return AppLayerResult::ok();
+    }
+
+    fn parse_response(&mut self, input: &[u8]) -> AppLayerResult {
+        // We're not interested in empty responses.
+        if input.is_empty() {
+            return AppLayerResult::ok();
+        }
+
+        if self.response_gap {
+            if !probe(input) {
+                // The parser now needs to decide what to do as we are not in sync.
+                // For this stun, we'll just try again next time.
+                return AppLayerResult::ok();
+            }
+
+            // It looks like we're in sync with a message header, clear gap
+            // state and keep parsing.
+            self.response_gap = false;
+        }
+        let mut start = input;
+        while !start.is_empty() {
```

Comment:
Over UDP ? we can have multiple PDUs in one packet ?

Reply:
This is (shameful) left-over from using the template to kickstart this development...